controllers/azurecluster_reconciler.go

```diff
@@ -152,46 +102,14 @@ func (s *azureClusterService) Delete(ctx context.Context) error {
 	ctx, _, done := tele.StartSpanWithLogger(ctx, "controllers.azureClusterService.Delete")
 	defer done()
 
-	if err := s.groupsSvc.Delete(ctx); err != nil {
+	if err := s.services[0].Delete(ctx); err != nil {
 		if errors.Is(err, azure.ErrNotOwned) {
-			if err := s.bastionSvc.Delete(ctx); err != nil {
-				return errors.Wrap(err, "failed to delete bastion")
-			}
-
-			if err := s.privateDNSSvc.Delete(ctx); err != nil {
-				return errors.Wrap(err, "failed to delete private dns")
-			}
-
-			if err := s.loadBalancerSvc.Delete(ctx); err != nil {
-				return errors.Wrap(err, "failed to delete load balancer")
-			}
-
-			if err := s.peeringsSvc.Delete(ctx); err != nil {
-				return errors.Wrap(err, "failed to delete peerings")
-			}
-
-			if err := s.subnetsSvc.Delete(ctx); err != nil {
-				return errors.Wrap(err, "failed to delete subnet")
-			}
-
-			if err := s.natGatewaySvc.Delete(ctx); err != nil {
-				return errors.Wrapf(err, "failed to delete NAT gateway")
-			}
-
-			if err := s.publicIPSvc.Delete(ctx); err != nil {
-				return errors.Wrapf(err, "failed to delete public IP")
-			}
-
-			if err := s.routeTableSvc.Delete(ctx); err != nil {
-				return errors.Wrap(err, "failed to delete route table")
-			}
-
-			if err := s.securityGroupSvc.Delete(ctx); err != nil {
-				return errors.Wrap(err, "failed to delete network security group")
-			}
-
-			if err := s.vnetSvc.Delete(ctx); err != nil {
-				return errors.Wrap(err, "failed to delete virtual network")
+			// If the resource group is not managed we need to delete resources inside the group one by one.
+			// services are deleted in reverse order from the order in which they are reconciled.
+			for i := len(s.services) - 1; i >= 1; i-- {
+				if err := s.services[i].Delete(ctx); err != nil {
+					return errors.Wrap(err, "failed to delete AzureCluster service")
```

Comment:
Can we access the `serviceName` variable from `s.services[i]`? That way we could do something like `errors.Wrap(err, fmt.Sprintf("failed to delete AzureCluster service '%s'", serviceName)`.

Reply:
not currently. Each service in `s.services` is a `azure.Reconciler` which only implements `Reconcile()` and `Delete()`. I actually thought about this as well and considered adding `ServiceName()` to the interface so we could do this but eventually decided not to pursue that route since it requires modifying every single service and the error we get back already contains the service name/type.

Reply:
Where does the error add the service name? It looks like we add the `serviceName` in the errors coming out of `CreateResource()` and `DeleteResource()` in async.go, but if we just return an error like `errors.Errorf("%T is not a network.Subnet", result)` in the reconciler I'm not sure if that will include the service name.

Reply:
this should be solved now, thanks for pointing that out